inc/class-core-sitemaps-provider.php

```diff
@@ -99,4 +92,75 @@ public function get_url_list( $page_num ) {
 	public function rewrite_query() {
 		return 'index.php?sitemap=' . $this->slug . '&paged=$matches[1]';
 	}
+
+	/**
+	 * Return object type being queried.
+	 *
+	 * @return string Name of the object type.
+	 */
+	public function get_queried_type() {
+		$type = $this->sub_type;
+		if ( empty( $type ) ) {
+			$type = $this->object_type;
+		}
+
+		return $type;
+	}
+
+	/**
+	 * Query for determining the number of pages.
+	 *
+	 * @param string $type Object Type.
+	 * @return int Total number of pages.
+	 */
+	public function max_num_pages( $type = null ) {
+		if ( empty( $type ) ) {
+			$type = $this->get_queried_type();
+		}
+		$query = new WP_Query(
+			array(
+				'fields'         => 'ids',
+				'orderby'        => 'ID',
+				'order'          => 'ASC',
+				'post_type'      => $type,
+				'posts_per_page' => CORE_SITEMAPS_POSTS_PER_PAGE,
+				'paged'          => 1,
+			)
+		);
+
+		return isset( $query->max_num_pages ) ? $query->max_num_pages : 1;
+	}
+
+	/**
+	 * List of sitemaps exposed by this provider.
+	 *
+	 * @return array List of sitemaps.
+	 */
+	public function get_sitemaps() {
+		$sitemaps = array();
+
+		foreach ( $this->get_object_sub_types() as $type ) {
+			$total = $this->max_num_pages( $type->name );
+			for ( $i = 1; $i <= $total; $i ++ ) {
+				$slug       = implode( '-', array_filter( array( $this->slug, $type->name, (string) $i ) ) );
+				$sitemaps[] = $slug;
+			}
+		}
+
+		return $sitemaps;
+	}
+
+	/**
+	 * Stub a fake object type, to get the name of.
+	 * This attempts compatibility with object types such as post, category, user.
+	 * This must support providers for multiple sub-types, so a list is returned.
+	 *
+	 * @return array List of object types.
+	 */
+	public function get_object_sub_types() {
+		$c       = new stdClass();
```

Comment:
This is a smart hack, but still seems like a hack. Would be more robust if this returned an empty array and anywhere that this function is used, we add a sanity check to make sure the array is not empty before looping through the array (which could become needed if someone filtered all of their taxonomies to private for some reason, for example).

Reply:
Simplified and abstracted this more cleanly, now providers with < 2 sub types are handled cleanly.